src/treemodel/schema.js

```diff
@@ -0,0 +1,417 @@
+	 *
+	 * @param {String} type Paths' type. Possible values are `ALLOW` or `DISALLOW`.
+	 * @param {String} [attribute] If set, only paths registered for given attribute will be returned.
+	 * @returns {Array} Paths registered in the item.
+	 * @private
+	 */
+	_getPaths( type, attribute ) {
+		const source = type === 'ALLOW' ? this._allowed : this._disallowed;
+		const paths = [];
+
+		for ( let item of source ) {
+			if ( item.attribute === attribute ) {
+				paths.push( item.path.slice() );
+			}
+		}
+
+		return paths;
+	}
+
+	/**
+	 * Checks whether this item has any registered path of given type that matches provided path.
+	 *
+	 * @param {String} type Paths' type. Possible values are `ALLOW` or `DISALLOW`.
+	 * @param {Array.<String>} checkPath Path to check.
+	 * @param {String} [attribute] If set, only paths registered for given attribute will be returned.
+	 * @returns {Boolean} `true` if item has any registered matching path, `false` otherwise.
+	 * @private
+	 */
+	_hasMatchingPath( type, checkPath, attribute ) {
+		const itemPaths = this._getPaths( type, attribute );
+
+		checkPath = checkPath.slice();
+
+		// We check every path registered (possibly with given attribute) in the item.
+		for ( let itemPath of itemPaths ) {
+			// We have one of paths registered in the item.
+
+			// Now we have to check every item name from the path to check.
+			for ( let checkName of checkPath ) {
+				// Every item name is expanded to all names of items that item is extending.
+				// So, if on item path, there is an item that is extended by item from checked path, it will
+				// also be treated as matching.
+				const chain = this._schema._extensionChains.get( checkName );
+
+				// Since our paths have to match in given order, we always check against first item from item path.
+				// So, if item path is: B D E
+				// And checked path is: A B C D E
+				// It will be matching (A won't match, B will match, C won't match, D and E will match)
+				if ( chain.indexOf( itemPath[ 0 ] ) > -1 ) {
+					// Every time we have a match, we remove it from `itemPath` so we can still check against first item.
+					itemPath.shift();
+				}
+			}
+
+			// If `itemPath` has no items it means that we removed all of them, so we matched all of them.
+			// This means that we found a matching path.
+			if ( itemPath.length === 0 ) {
+				return true;
+			}
+		}
+
+		// No matching path found.
+		return false;
+	}
+
+	/**
+	 * Custom toJSON method to solve child-parent circular dependencies.
+	 *
+	 * @returns {Object} Clone of this object with the parent property replaced with its name.
+	 */
+	toJSON() {
+		const json = clone( this );
+
+		// Due to circular references we need to remove parent reference.
+		json._schema = '[treeModel.Schema]';
+
+		return json;
+	}
+}
+
+/**
+ * Schema is a run-time created and modified description of which entity in Tree Model is allowed to be inside another
```

Comment:
This class needs a more high-level introduction to be understandable. Something like:

Schema is a definition of the structure of the document. It allows to define which tree model items (element, text, etc.) can or be nested within which and which attributes can be applied to them. It's created during the run-time of the application, typically by features. Also, the features can query the schema to learn what's the allowed structure and act accordingly.

For instance, if a feature wants to define that an attribute `bold` is allowed on the text it needs to register this rule like this:

```
editor.document.schema.allow( '$text', 'bold' );
```

Note: items prefixed with `$` are special group of items. `$text` means text nodes and text-like content. `$block` means block elements.

When registering an item it's possible to tell that this item should inherit from some other existing item. E.g. `p` can inherit from `$block`, so when an attribute `center` is allowed on the `$block` it will automatically be also allowed on the `p` element.

etc. etc. Try to write this from the user perspective. Don't document the implementation, but the feature.